scheduler/src/ratelimit/token_bucket_filter.clj

```diff
@@ -0,0 +1,88 @@
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns ratelimit.token-bucket-filter)
+
+(defn create-tbf
+  "Make a token bucket filter (tbf), with a given refill rate (units per timestep), a maximum token capacity, and give it the
+  current timestamp.
+
+  This token bucket filter is a bit unusual, instead of earning tokens that you can later spend, we allow the current
+  tokens to be negative. Thus, you can (in one request) spend an unbounded number of tokens. However, you will be
+  in debt until your token balance is at least 0. This design allows the tokens in a single request to exceed the max-tokens
+  in the tbf. (If we didn't do it this way, max-tokens is an implicit maximum limit on request size.) This way
+  we can independently pick the burst size in the tbf and the maximum request size.
+
+  This library does not handle blocking or other operations. Typical workflows are to:
+
+  ** Admission control **
+  This is good when you don't know how big a request will be until afterwards.
+
+  Use in-debt? as part of request admission control, rejecting a request if it returns true. When a request completes (or is aborted),
+  take the tokens that it used, affecting the next request.
+
+  ** Delaying requests **
+  If you want to delay requests instead of rejecting. Take the tokens the request needs, pause for the returned pause time.
+  (E.g., schedule the message to be sent at that particular pause time.
+  "
+  ([^double token-rate ^long max-tokens ^long current-time ^long current-tokens]
+   {:pre [(> token-rate 0)
+          (>= max-tokens 0)
+          (>= current-time 0)]}
+   {:token-rate token-rate :max-tokens max-tokens :last-update current-time :current-tokens current-tokens})
+  ([^double token-rate ^long max-tokens ^long current-time]
+   (create-tbf token-rate max-tokens current-time 0)))
+
+(defn- add-tokens-helper
+  "Add tokens (if there are any). If less than a whole token added, keep the current state. Returns the current state.
+  Enforces max-tokens limit. Allows tokens to be removed or added. If no timestamp supplied, just alters the token count."
+  ([{:keys [current-tokens max-tokens] :as tbf} ^long tokens-to-dispense ^long current-time]
+   (if (zero? tokens-to-dispense)
+     ; Only dispense when we have at least a whole token to add on.
+     tbf
+     (-> tbf
+         (assoc :last-update current-time)
+         (update :current-tokens #(-> % (+ tokens-to-dispense) (min max-tokens))))))
+  ([tbf ^long tokens-to-dispense]
```

Comment:
I think this overload can be avoided and the caller can pass in the `current-time` argument explicitly.

Reply:
That line of code has to exist, either here, or in the caller (spend-tokens) when I want to update the token count without updating the last-updated.

Reply:
Yes, that is what I requested - for the caller to pass it in explicitly.